### ISSUE TITLE
a reasonable :backward behavior

### ISSUE THREAD

**Issue by soumith** [torch/nn#262]
@koraykv mentioned that implementing one-shot backward in #234 broke some of their internal tests.

I guess there was in implicit contract since many years that :backward is :updateGradInput + :accGradParameters. 

We do want a direct :backward for performance reasons, whether it is called :backward or something else. :backward will be the absolute most natural naming.

Now the question is, how hard is it to fix your internal tests/modules to have a direct :backward? 
Depending on your answer, we can figure out something else, that will really be a worse-off solution in terms of core-design.

**Comment by dominikgrewe** [comment]
Picking this up again as we haven't found a good solution for this.

As you mentioned, Soumith, there's an understanding that `backward` always calls `updateGradInput` and `accGradParameters`. It is also advised in the [docs](https://github.com/torch/nn/blob/master/doc/module.md#gradinput-backwardinput-gradoutput) to not override the `backward` function.

A core module such as `nn.Sequential` really shouldn't violate this contract, as it can lead to subtle bugs that are hard to track down. If it's really necessary to have a one-shot backwards pass it may be better to make this explicit, e.g. by deriving a class from `nn.Sequential` with an appropriate name and clearly documenting its behaviour. Any thoughts?

**Comment by hughperkins** [comment]
Could updateGradInputs do all the work, and accGradParameters becomes a `nop`?

**Comment by dominikgrewe** [comment]
@hughperkins I don't quite understand your comment. Are you suggesting to do this for all modules?

**Comment by hughperkins** [comment]
no, I mean, instead of making backward shortcircuit updateGradInputs and accGradParameters, instead do, just for the modules which one-shot do both updategrad and updateparams at the same time:

(hmmm, and my lua is getting a bit rusty, might look python-like :-P)
```
class MyOneshotModule(object):
    def backward(self, input, gradOutput):
       self.updateGradInputs(input, gradOutput)
       self.updateaccGradParameters(input, gradOutput)

    def updateGradInputs(self, input, gradOutput):
       (self.gradInput, self.paramChanges) = self._oneShotDoBackwards()
       return self.gradInput

    def accGradParameters(self, input, gradOutput):
        self.weights = self.weights:add(self.paramChanges)
```
This meets the requireemtns of running the code that calculates the gradInput, and the parameterChanges in one-shot, whilst being basically compliant with the current api.  The only violation would be for timing purposes, since updateGradInputs will run slowly, and accGradParameters will be essentially instantaneous.

**Comment by dominikgrewe** [comment]
I see. You could even lazily call `self._oneShotBackwards()` in both `updateGradInputs` and `accGradParameters` if you haven't done so for the given `input` and `gradOutput`. Then it would be safe, regardless of how you use it. You'd need a way to hash tensors though and this adds overhead of course.
But it'd be a cleaner solution imo.

**Comment by hughperkins** [comment]
You could, though it might be simpler to simply state in the contract that calling `accGradParameters` prior to `updateGradInputs` is a violation, with undefined behavior?

**Comment by hughperkins** [comment]
(Hmmm, or, since calling updateOutput is alread in the contract, as needing to be called first, simply get it to set a flag saying it's been called, and then check for this in the others:
```
function MyOneShot.updateOutput(self, input)
    self._gradientDirty = true
end
function MyOneShot.updateGradInput(self, input, gradOutput)
    if self._gradientDirty then
        (self.gradInput, self.paramsChanges) = self._oneShotBackwards(self, input, gradOutput)
        self._gradientDirty = false
    end
    return self.gradInput
end
function MyOneShot.accParams(self, input, gradOutput)
    if self._gradientDirty then
        self._oneShotBackwards(self, input, gradOutput)
        self._gradientDirty = false
    end
end
```

**Comment by dominikgrewe** [comment]
But that only guarantees that the input is the same. gradOutput may still change I think.
@soumith Do you have any thoughts on this?

**Comment by soumith** [comment]
I've thought of doing this for batchnorm, where accgradparams checks if updateGradInput is already executed for the given input/gradOutput and reuse the results. 

Hashing a tensor is not a small amount of computation, but in some cases such as batchnorm it might outweigh the benefits that come with being able to do so.

Last time we discussed tensor hashes, it did not go anywhere right? https://github.com/torch/torch7/pull/119

**Comment by hughperkins** [comment]
Example of where input stays same but gradOutput changes?

**Comment by dominikgrewe** [comment]
I don't have an example, but it's allowed, isn't it? :)

**Comment by hughperkins** [comment]
Given how 'unclean' and slow hashing the tensors sound to me, I think it would be good to have a specific use-case in mind for why this is necessary, or else make it illegal in the contract.

**Comment by nicholas-leonard** [comment]
Not sure where this is going, but I think we should just create a new method that calls backward by default.
```lua
function Module:backprop(input, gradInput, scale)
   return self:backward()
end
```
And then we only have to define it for a few existing containers :
```lua
funciton Sequential:backprop(input, gradInput, scale)
   for i in ...
      module:backprop(input[i], ...)
   end
end
```
So that way its backward compatible and doesn't break anything. But maybe I missed something, or don't sufficiently understand the problem...

**Comment by fmassa** [comment]
@soumith @dominikgrewe did you guys reached an agreement wrt the `:backward` method ? It would allow not only faster training, but also to [reduce memory usage during training](https://github.com/soumith/dcgan.torch/issues/8#issuecomment-169580958)

**Comment by hughperkins** [comment]
Is there any reason why we would ever want to call `accGradParameters`, without first calling `updateGradInput`?  If not, simply stating in the contract (which is ... where? :-P ) that it calling `accGradParameters` without first calling `updateGradInput` is illegal seems to solve the problem?

Then, just make `updateGradInput` do all the work, and have `accGradParameters` do little or nothing.

**Comment by fmassa** [comment]
@hughperkins just for information, check the comment from @koraykv here https://github.com/torch/nn/pull/234#issuecomment-98099276

**Comment by hughperkins** [comment]
@fmassa ok, but that comment is irrelevant in the case of one-shot, since there is no time-gain in not calculating the weight gradients: they're calculated 'for free' anyway, which is why we're thinking through such modules.

**Comment by hughperkins** [comment]
Oh, you mean, we actually have two ways of calling each module?  Like:

```
:backward()  -- => calculates both gradInput and gradWeight.  Does it one-shot, but not totally speedy
```
```
:updateGradInput()  -- => calculates only gradInput.  Faster than calling :backward(), in our hypothetical scenario
```

**Comment by fmassa** [comment]
@hughperkins precisely. In cases like [Neural Style](https://github.com/jcjohnson/neural-style/blob/master/neural_style.lua#L320) or [DeepDream](https://github.com/eladhoffer/DeepDream.torch/pull/2) for example, it's much faster to just compute the grads wrt the input.

**Comment by hughperkins** [comment]
Alright, so, in that case, in such modules, it seems we would have three methods?

`backward()` => does the one-shot, calculates gradInput, gradWeights; doesnt have to call `updateGradInput` etc, though it can, if it feels like it, couleur du ciel...
`updateGradInput()` => mandatory method.  updates `gradInput`
`accGradParameters()` => mandatory method, *if* the class has parameters to update.  updates module internal parameter gradients, typically `self.gradWeight`, `self.gradBias`

How's that sound?  Seems like simple, and avoids messing around with contracts, and fits all use cases?

**Comment by fmassa** [comment]
@hughperkins but some modules have optimized version for the `backward`, which is faster than computing `updateGradInput` and `accGradParameters` sequentially, see for example [SpatialBatchNormalization](https://github.com/torch/nn/blob/master/SpatialBatchNormalization.lua#L120-L122)

**Comment by fmassa** [comment]
@hughperkins yes, that is the original plan. But it created some problems with internal tests in DeepMind, that's why we created this issue.

**Comment by hughperkins** [comment]
I think the test problem was precisely because `updateGradInput` and/or `accGradParameters` were entirely removed?

**Comment by fmassa** [comment]
@hughperkins not really, check https://github.com/torch/nn/pull/234

**Comment by hughperkins** [comment]
@fmassa Hmmm

@koraykv Seems like https://github.com/torch/nn/pull/234/files  doesnt actually remove the old `updateGradInput` or `accGradParameters` methods, so they should continue to work ok?  What was concretely breaking in the tests?  Can you provide a few lines of test code to reproduce the issue?

**Comment by hughperkins** [comment]
Although actually @dominikgrewe 's suggestion, right at the head of this thread, makes a lot of sense.  Otherwise it's confusing having two different sets of implementations inside the same module perhaps?

"there's an understanding that backward always calls updateGradInput and accGradParameters. It is also advised in the docs to not override the backward function.

"A core module such as nn.Sequential really shouldn't violate this contract, as it can lead to subtle bugs that are hard to track down. If it's really necessary to have a one-shot backwards pass it may be better to make this explicit, e.g. by deriving a class from nn.Sequential with an appropriate name and clearly documenting its behaviour."
